File: src/main/java/gg/mineral/practice/util/messages/ChatMessage.java
package gg.mineral.practice.util.messages;

import org.bukkit.entity.Player;

public class ChatMessage extends Message {
	String addition;

	ChatMessage(String s) {
		message = s;
	}

	public ChatMessage(String s, String c, boolean bold) {
		message = s;
		formatMessage(c, bold);
	}

	public ChatMessage(String s, String c) {
		message = s;
		formatMessage(c, false);
	}

	protected void formatMessage(String c, boolean bold) {
		this.addition = bold ? c + CC.B : c;
		message = addition + message;
	}

	public ChatMessage highlightText(String c, String... highlighted) {

		for (String s : highlighted) {
			message = message.replace(s, c + s + this.addition);
		}

		return this;
	}

	public ChatMessage replace(String message, String replacement) {
		this.message = this.message.replace(message, replacement);
		return this;
	}

	public ChatMessage clone() {
		return new ChatMessage(message);
	}

	@Override
	public void send(Player p) {
		p.sendMessage(message);
		return;
	}
}

File: src/main/kotlin/gg/mineral/practice/util/messages/ChatMessage.kt
package gg.mineral.practice.util.messages

open class ChatMessage(
    string: String,
    val colorPrefix: String = "",
    val bold: Boolean = false
) :
    Message(string) {
    open val prefix by lazy { if (bold) colorPrefix + CC.B else colorPrefix }

    open fun highlightText(c: String, vararg highlighted: String): ChatMessage {
        for (s in highlighted) {
            var index = messageBuilder.indexOf(s)
            while (index != -1) {
                messageBuilder.replace(index, index + s.length, c + s + this.prefix)
                index = messageBuilder.indexOf(s, index + (c + s + this.prefix).length)
            }
        }
        return this
    }

    open fun replace(target: String, replacement: String): ChatMessage {
        var index = messageBuilder.indexOf(target)
        if (index == -1) throw AssertionError("Target string not found in the message.")
        while (index != -1) {
            messageBuilder.replace(index, index + target.length, replacement)
            index = messageBuilder.indexOf(target, index + replacement.length)
        }
        return this
    }

    override fun prepend(builder: StringBuilder): StringBuilder = builder.insert(0, this.prefix)

    override fun clone() = ChatMessage(messageBuilder.toString(), this.colorPrefix, bold)
}
